crates/krator/src/state.rs

```diff
@@ -0,0 +1,200 @@
+
+/// Guard for preventing manual construction on Transition::Next.
+pub struct StateHolder<S: ResourceState> {
+    pub(crate) state: Box<dyn State<S>>,
+}
+
+impl<S: ResourceState> From<StateHolder<S>> for Box<dyn State<S>> {
+    fn from(holder: StateHolder<S>) -> Box<dyn State<S>> {
+        holder.state
+    }
+}
+
+/// Represents result of state execution and which state to transition to next.
+pub enum Transition<S: ResourceState> {
+    /// Transition to new state.
+    Next(StateHolder<S>),
+    /// Stop executing the state machine and report the result of the execution.
+    Complete(anyhow::Result<()>),
+}
+
+/// Mark an edge exists between two states.
+pub trait TransitionTo<S> {}
+
+impl<S: ResourceState> Transition<S> {
+    // This prevents user from having to box everything AND allows us to enforce edge constraint.
+    /// Construct Transition::Next from old state and new state. Both states must be State<PodState>
+    /// with matching PodState. Input state must implement TransitionTo<OutputState>, which can be
+    /// done manually or with the `TransitionTo` derive macro (requires the `derive` feature to be
+    /// enabled)
+    #[allow(clippy::boxed_local)]
+    pub fn next<I: State<S>, O: State<S>>(_i: Box<I>, o: O) -> Transition<S>
+    where
+        I: TransitionTo<O>,
+    {
+        Transition::Next(StateHolder { state: Box::new(o) })
+    }
+
+    /// Represents a transition to a new state that is not checked against the
+    /// set of permissible transitions. This is intended only for use by generic
+    /// states which cannot declare an exit transition to an associated state
+    /// without encountering a "conflicting implementations" compiler error.
+    #[allow(clippy::boxed_local)]
+    pub fn next_unchecked<I: State<S>, O: State<S>>(_i: Box<I>, o: O) -> Transition<S> {
+        Transition::Next(StateHolder { state: Box::new(o) })
+    }
+}
+
+/// Convenience redefinition of Arc<RwLock<T>>
+pub type SharedState<T> = std::sync::Arc<tokio::sync::RwLock<T>>;
+
+#[async_trait::async_trait]
+/// A trait representing a node in the state graph.
+pub trait State<S: ResourceState>: Sync + Send + 'static + std::fmt::Debug {
+    /// Provider supplies method to be executed when in this state.
+    async fn next(
+        self: Box<Self>,
+        shared: SharedState<S::SharedState>,
+        state: &mut S,
+        manifest: &S::Manifest,
+    ) -> Transition<S>;
+
+    /// Provider supplies JSON status patch to apply when entering this state.
+    async fn status(&self, state: &mut S, manifest: &S::Manifest) -> anyhow::Result<S::Status>;
+}
+
+/// Iteratively evaluate state machine until it returns Complete.
+pub async fn run_to_completion<S: ResourceState>(
+    client: &kube::Client,
+    state: impl State<S>,
+    shared: SharedState<S::SharedState>,
+    object_state: &mut S,
+    manifest: Arc<RwLock<S::Manifest>>,
+) where
+    S::Manifest: Resource + Meta + DeserializeOwned,
```

Comment:
This would require that the container state machine has a type which implements `Resource` and `Meta`. I think the state machine API should be decoupled from Kubernetes objects, and then if you want to use `Operator` and this `run_to_completion` implementation, you need these additional traits.

Reply:
Wouldn't a generic state machine also require a `run_to_completion` method?

Reply:
I'm a little confused. So this is method is generic for any k8s API resource. The traits for the state machine are designed such that you can implement a state machine for non API resources, but writing the `run_to_completion` method is left to the user in this case (which is what we did for Containers).

Reply:
Oh right...I forgot that is left as an exercise to the user